### ISSUE TITLE
[Support] Skewed Output

### ISSUE THREAD

**Issue by tejasftc** [johnboiles/obs-mac-virtualcam#160]
**Describe the issue**
When I run the OBS Virtual Camera for Mac, I get some weird output.

**To Reproduce**
1. I downloaded the package with the latest version
2. I went to OBS and selected "Start Virtual Camera"
3. I went to Google Meet and switched my camera input to the virtual camera

**Behavior**
The virtual camera is showing a weird image in Google Meet.

**Screenshots**

**Desktop (please complete the following information):**
 - OS version: 10.15.5
<img width="1440" alt="Screen Shot 2020-06-03 at 11 29 20 AM" src="https://user-images.githubusercontent.com/43117594/83677381-f014f380-a590-11ea-82d5-3b45fc88c646.png">
<img width="1440" alt="Screen Shot 2020-06-03 at 11 30 18 AM" src="https://user-images.githubusercontent.com/43117594/83677391-f4d9a780-a590-11ea-88e2-c71342bb4424.png">

 - Application (if applicable): Chrome
 - Virtualcam version: 1.2.0
 - OBS version: 25.0.8

**Additional context**

**Comment by gxalpha** [comment]
Uhhmm... That should have been fixed 😄 
Are you sure you are running 1.2.0?
If so, it should show "1.2.0" at the bottom of the test card. Please report back if it does.
<img width="1280" alt="Bildschirmfoto 2020-06-03 um 21 33 11" src="https://user-images.githubusercontent.com/59806498/83680915-e798d900-a5e1-11ea-912a-95c31056fac1.png">

**Comment by tejasftc** [comment]
I don't see the test card either since it is also showing a skewed image. How do I totally uninstall it totally? I tried uninstalling using what was online but I am wondering if I am missing something. Here's a screenshot of what my test card looks like: 
![Screen Shot 2020-06-03 at 1 06 48 PM](https://user-images.githubusercontent.com/43117594/83684003-26577080-a59b-11ea-9f04-caaccbf657c9.png)

**Comment by tejasftc** [comment]
> I don't see the test card either since it is also showing a skewed image. How do I totally uninstall it totally? I tried uninstalling using what was online but I am wondering if I am missing something. Here's a screenshot of what my test card looks like:
> ![Screen Shot 2020-06-03 at 1 06 48 PM](https://user-images.githubusercontent.com/43117594/83684003-26577080-a59b-11ea-9f04-caaccbf657c9.png)

Is there another way to check the version?

**Comment by gxalpha** [comment]
Insert this command into Terminal.app:
`cat /Library/CoreMediaIO/Plug-Ins/DAL/obs-mac-virtualcam.plugin/Contents/Info.plist | grep -a3 CFBundleShortVersionString` and post the results. It should say the version number in there

**Comment by tejasftc** [comment]
![image](https://user-images.githubusercontent.com/43117594/83691923-daf78f00-a5a7-11ea-9e91-63c530c37770.png)
Here’s what I got

**Comment by gxalpha** [comment]
Have you rebooted your mac?

**Comment by tejasftc** [comment]
Yes I tried that.

**Comment by gxalpha** [comment]
Could you attach an OBS log of when the problem occurred? That would help understand the problem

**Comment by tejasftc** [comment]
Yeah sure. Here it is:
[2020-06-03 13-06-27.txt](https://github.com/johnboiles/obs-mac-virtualcam/files/4727161/2020-06-03.13-06-27.txt)

**Comment by johnboiles** [comment]
Very interesting. I thought we had this solved. I seel

```
13:06:27.539: 	base resolution:   1440x900
13:06:27.539: 	output resolution: 860x538
```

If you change both resolutions to 1280x720 does that fix things?

![image](https://user-images.githubusercontent.com/218876/83807559-240f1800-a668-11ea-8353-f6165feda0bc.png)

To be clear, any resolution _should_ work. So this is definitely a bug.

**Comment by tejasftc** [comment]
> Very interesting. I thought we had this solved. I see this in the logs
> 
> ```
> 13:06:27.539: 	base resolution:   1440x900
> 13:06:27.539: 	output resolution: 860x538
> ```
> 
> If you change both resolutions to 1280x720 does that fix things?
> 
> ![image](https://user-images.githubusercontent.com/218876/83807559-240f1800-a668-11ea-8353-f6165feda0bc.png)
> 
> To be clear, any resolution _should_ work. So this is definitely a bug.

I tried changing the resolution as well. Here's what I got:
<img width="1424" alt="Screen Shot 2020-06-04 at 2 40 17 PM" src="https://user-images.githubusercontent.com/43117594/83813030-6db03080-a671-11ea-8d9f-58e9b2f4c1c0.png">
[2020-06-04 14-38-13.txt](https://github.com/johnboiles/obs-mac-virtualcam/files/4732866/2020-06-04.14-38-13.txt)

Thanks for your help by the way.

**Comment by johnboiles** [comment]
Hmm. I also see 

```14:38:13.948: NV12 texture support not available```

Could you try using a different format (I think this is in the advanced settings page iirc.

**Comment by tejasftc** [comment]
It worked! Thanks a lot for your help!
<img width="1171" alt="Screen Shot 2020-06-04 at 3 44 26 PM" src="https://user-images.githubusercontent.com/43117594/83817741-68a3af00-a67a-11ea-8c29-0e962a90c0e4.png">

**Comment by johnboiles** [comment]
Very interesting I wonder if this is an obs bug or if I'm doing something wrong with converting pixel formats. @dodgepong have you seen anything like this in other issues on OBS? Seems like @tejasftc had the format set to NV12 even though his computer reported that it wasn't supported. This then seemed to break the pixel format conversion to yuv422 that my plugin does

**Comment by dodgepong** [comment]
I've never heard of any issue like this before, no.

**Comment by jp9000** [comment]
The `NV12 texture not supported` message is for whether full NV12 textures are supported, which is a Windows-specific thing at the moment.  It doesn't affect OBS' usage of NV12.  OBS will just use shaders to implement NV12 if NV12 textures aren't available, and if there was a bug with that, then you would see it in all outputs because it's the default format.

For format conversion, we usually use FFmpeg.  So that shouldn't be in our territory either.

When the base and output resolutions don't match, it uses the GPU to scale, but there isn't any known issue with that either, otherwise it would affect all outputs and not just virtualcam.

Without having a full grasp on all your code I can't really say anything substantive, but I don't *think* OBS would be giving bad NV12 data, because NV12 is a very high frequency code path (it's the default).  It would also produce on Windows in that case.  Personally, I'd just check the memory of the data at all parts of the pipeline after you download it from the GPU to figure out where it's coming from.  I'd recommend using a test image that would make comparing scanlines of the image more simple when checking for errors.

It looks like a bad linesize/pitch value somewhere.

**Comment by tejasftc** [comment]
Do the other formats cause the output use more CPU Power? It really slowed down my computer so I couldn’t really use it.

**Comment by gxalpha** [comment]
OBS takes a lot of power in general, I'm afraid a dual core i5 processor (according to your log file) may not be enough to do lot of things simultaneously.
I'd like to put your log into the OBS analyzer and look if it is a problem, but I don't know how to :D

**Comment by jp9000** [comment]
NV12 probably uses the least amount of CPU processing.  It uses the least data.

**Comment by tejasftc** [comment]
Ok. Thanks
Tejas Thiyagarajan
tejas.thiyagarajan@gmail.com

**Comment by johnboiles** [comment]
@jp9000 thank you for chiming in! [This](https://github.com/johnboiles/obs-mac-virtualcam/blob/master/src/obs-plugin/plugin-main.mm#L49-L58) is the whole of my pixel format conversion code in the plugin:

```
    obs_get_video_info(&videoInfo);
    struct video_scale_info conversion = {};
    conversion.format = VIDEO_FORMAT_UYVY;
    conversion.width = videoInfo.output_width;
    conversion.height = videoInfo.output_height;
    obs_output_set_video_conversion(output, &conversion);
    if (!obs_output_begin_data_capture(output, 0)) {
        return false;
    }
```

That should make it so that the output's `raw_video` method _always_ gets a framebuffer with the pixel format `VIDEO_FORMAT_UYVY` right? Is it correct to say that this _shouldn't_ depend on the current 'Color Format' setting? I'm trying to understand what about the OBS 'Color Format' setting could mess up the framebuffer I receive since I have this conversion in place.

**Comment by jp9000** [comment]
Are you able to reproduce this @johnboiles?

**Comment by johnboiles** [comment]
Unfortunately no, so im flying blind

**Comment by jp9000** [comment]
Hm, I wonder what's going on.  At some point I'll test this myself and try to debug.  My guess is that the client application itself is using the wrong format or seeing it as the wrong format.  I'm just kind of overburdened with my own stuff at the moment (and let's not forget the 115 pull requests on main) so I don't really feel comfortable spending time on this right at the moment.

By the way, is there a reason why you're converting to UYVY rather than just using NV12?

**Comment by gxalpha** [comment]
@jp9000 to my knowledge UYVY is used because it worked and others didn't, but that was back when this was just a proof of concept and it just stayed.

**Comment by johnboiles** [comment]
@jp9000 yes not urgent at all; this doesn't seem to be a widespread problem. Your time is probably best spent working on main! Feel free to drop off the conversation here and we won't be offended :)

What @gxalpha says is correct, UYVY was the first thing I was able to make work and I just ran with it. I spent a little bit of time investigating further in https://github.com/johnboiles/obs-mac-virtualcam/issues/102 and I don't think Apple's Core Media APIs support NV12 as a camera format. I also tried to get I444 to work (using `kCMVideoCodecType_AppleProRes4444`) but that didn't seem to work either. Perhaps RGB could be made to work. Obviously if I can match a pixel format that OBS supports natively with a format that works directly as a Core Media IO DAL plugin format then we can avoid the pixel format conversion which should significantly lower CPU usage I think. [This](https://developer.apple.com/documentation/coremedia/1564239-video_codec_constants?language=objc) is the list of formats I think I have to pick from on the CoreMedia side. It might also be possible to use something from this [list](https://developer.apple.com/documentation/coremedia/cmformatdescription/1564244-video_pixel_format_constants) of pixel formats though I'm not exactly sure how.

**Comment by TayKangSheng** [comment]
I am also facing the same problem. It only started maybe 2 weeks - 1 month back. 

1. This is supposed to be the test card
    <img width="774" alt="Screen Shot 2020-06-29 at 2 05 28 PM" src="https://user-images.githubusercontent.com/10722197/85979328-d42b3300-ba12-11ea-9492-282a2d11349a.png">

1. and this is when the virtual cam is started.
    <img width="778" alt="Screen Shot 2020-06-29 at 2 05 11 PM" src="https://user-images.githubusercontent.com/10722197/85979301-c7a6da80-ba12-11ea-8398-d53cd68b2b7c.png">

I am using a mac, but not sure how to get to the logs. if someone can let me know how to get to the logs, i would love to provided it here. 

I tried changing the resolution and also color format in advanced settings. Both didn't work. The warpping looks a bit different though.

**Comment by gxalpha** [comment]
@TayKangSheng There are two logs needed:
* The mac logs from the mac console: For this, open Console.app, search for `obs-mac-virtualcam`, reproduce the problem and check for any error outputs or just taker a screenshot and post it here
* The OBS logs: Restart OBS, reproduce the problem and then go to Help -> Log Files -> Upload current log files. Post the link you get here.

Also please insert `cat /Library/CoreMediaIO/Plug-Ins/DAL/obs-mac-virtualcam.plugin/Contents/Info.plist | grep -a1 CFBundleVersion` into Terminal.app and post the result here.

**Comment by mgw-sbic** [comment]
I'm having a similar thing, just trying it today. It was working fine initially. I wanted more virtual cams, so I duplicated OBS.app to OBS1.app and turned on both virtual cameras. Now, the video is always like the others. Even after stopping an restarting both OBS and Chrome. 
cat /Library/CoreMediaIO/Plug-Ins/DAL/obs-mac-virtualcam.plugin/Contents/Info.plist | grep -a1 CFBundleVersion
	<string>1.2.0</string>
	<key>CFBundleVersion</key>
	<string>1.2.0</string>

**Comment by gxalpha** [comment]
Due to the way the virtual camera works, only one virtual camera can be used.

Could you try uninstalling and then reinstalling the virtual camera? (https://github.com/johnboiles/obs-mac-virtualcam/wiki/Installing)

**Comment by mgw-sbic** [comment]
> Due to the way the virtual camera works, only one virtual camera can be used.
> 
> Could you try uninstalling and then reinstalling the virtual camera? (https://github.com/johnboiles/obs-mac-virtualcam/wiki/Installing)

I didn't know how to uninstall so I just reinstalled, with no effect.

**Comment by gxalpha** [comment]
Just reinstalling often times does nothing with the DAL-Plugin until you reboot your Mac. The uninstalling tutorial is here: https://github.com/johnboiles/obs-mac-virtualcam/wiki/Installing#uninstalling

**Comment by mgw-sbic** [comment]
OK I uninstalled (thanks for the help!) and reinstalled, the title card is skewed as the others:
![Screen Shot 2020-06-30 at 2 37 24 PM](https://user-images.githubusercontent.com/24482481/86179618-71ec3400-badf-11ea-8e74-20631b564726.png)


One of the system.log entries:
Jun 30 14:07:55 Michaels-Mac-Pro com.apple.xpc.launchd[1] (com.apple.xpc.launchd.domain.user.501): Service "com.obsproject.obs-studio.28124" tried to register for endpoint "com.johnboiles.obs-mac-virtualcam.server" already registered by owner: com.apple.xpc.launchd.oneshot.0x10000005.obs

**Comment by mgw-sbic** [comment]
When I turn on the virtual camera I get:
![Screen Shot 2020-06-30 at 2 39 59 PM](https://user-images.githubusercontent.com/24482481/86179703-a3fd9600-badf-11ea-9260-16f38cee5765.png)

**Comment by TayKangSheng** [comment]
I didn't do anything.. but its fixed now 🤔 🤷

**Comment by jp9000** [comment]
@johnboiles I haven't looked at your code but make sure to check for race conditions.  "I didn't do anything but it magically fixed itself" and "I can't seem to reproduce" is a big sign of race conditions.  I initially had a race condition in my Windows implementation between the memory two processes would share and it caused some incorrect output.  Not saying that is the problem, but just throwing it out there as an item for your checklist.

**Comment by jp9000** [comment]
The output these peopel are reporting makes me think the sender/receiver processes have desynchronized dimension values.  Not entirely sure if that's the case but it's another thing for the checklist.

**Comment by jp9000** [comment]
For the people experiencing this problem:  see how changing OBS' resolution affects the skew, and report back.  Larger resolutions versus smaller resolutions.  You'd want to change the "Output/Scaled Resolution" in OBS' video settings to test that.

**Comment by johnboiles** [comment]
@TayKangSheng @mgw-sbic what version of macOS are y'all on? Can y'all attach the log from OBS (from the OBS help menu) as well?

@jp9000 thanks for thinking about this. Certainly does seem like the resolutions are somehow desynchronized, especially because the colors seem to be intact (which rules out pixel format differences I think). It's weird though because I send the resolution along with every frame, so I'm not sure how it could have gotten out of sync.

**Comment by saipanman95** [comment]
Hi All,
I just downloaded this plugin and it worked momentarily and now it doesn't. It is producing the same errors but it only happened after I closed down my google meet stream and attempted to reopen it.  It worked fine with video format initially now nothing.

**Comment by wesalvaro** [comment]
I changed the base resolution from 1920x1080 to 1080x1080 and it caused the skew.

**Comment by wesalvaro** [comment]
I opened Snap Camera and set its input to the OBS virtual camera...
This seems like it may be a clue: The snap camera preview accurately shows the demo card.

![Screen Shot 2020-07-09 at 12 36 35 PM](https://user-images.githubusercontent.com/13286/86994278-f5450e00-c1e0-11ea-9127-5ccb57ebd4db.png)

**Comment by wesalvaro** [comment]
Based on that, I restored the resolution to 1920x1080 and restarted.  
Hangouts in Chrome new accurately shows OBS.

**Comment by TayKangSheng** [comment]
@gxalpha Now I have this problem with one macbook and don't have this problem with my other macbook. I hope the details below help..

These are the logs from starting OBS + starting virtualcam. 
[console-obs.txt](https://github.com/johnboiles/obs-mac-virtualcam/files/4930698/console-obs.txt)

These are the logs when i am at google meet before starting virtualcam and after starting virtual cam.
[console-chrome.txt](https://github.com/johnboiles/obs-mac-virtualcam/files/4930697/console-chrome.txt)

This is what it looks like before virtualcam is started
<img width="921" alt="Screen Shot 2020-07-16 at 5 55 04 PM" src="https://user-images.githubusercontent.com/10722197/87657558-90178c80-c78d-11ea-819a-abff2412c205.png">

This is what it looks like after virtualcam is started
<img width="957" alt="Screen Shot 2020-07-16 at 5 55 10 PM" src="https://user-images.githubusercontent.com/10722197/87657574-973e9a80-c78d-11ea-8067-d3203c484026.png">

This is the output from `cat /Library/CoreMediaIO/Plug-Ins/DAL/obs-mac-virtualcam.plugin/Contents/Info.plist | grep -a1 CFBundleVersion`
```
	<string>1.2.0</string>
	<key>CFBundleVersion</key>
	<string>1.2.0</string>
```

This is the `Video` settings screen
<img width="993" alt="Screen Shot 2020-07-16 at 5 57 12 PM" src="https://user-images.githubusercontent.com/10722197/87657747-d240ce00-c78d-11ea-9e4f-5d28b343898c.png">

This is the `Advanced` settings screen
<img width="980" alt="Screen Shot 2020-07-16 at 5 57 06 PM" src="https://user-images.githubusercontent.com/10722197/87657766-d836af00-c78d-11ea-8b27-4f0bd91f6b5a.png">

This is my macOS version
<img width="603" alt="Screen Shot 2020-07-16 at 5 58 27 PM" src="https://user-images.githubusercontent.com/10722197/87657930-1338e280-c78e-11ea-95ab-a02bb680f20f.png">

**Comment by stevewww** [comment]
I had the same issue with the OBS virtual camera displaying in Google Meet (Chrome browser). I tried multiple resolution settings within OBS to change the behavior and a full uninstall and reinstall of both OBS and the virtual camera and rebooting the Mac. None of those steps resolved things. In the end I followed @wesalvaro lead and installed Snap Camera to experiment with a different application and it did not exhibit the issue at all. I then went back to test again in Google Meet and the skew problem was gone. I'll be on the look out for this resurfacing and better prepared to identify any triggers. HTH.

**Comment by truncj** [comment]
> I had the same issue with the OBS virtual camera displaying in Google Meet (Chrome browser). I tried multiple resolution settings within OBS to change the behavior and a full uninstall and reinstall of both OBS and the virtual camera and rebooting the Mac. None of those steps resolved things. In the end I followed @wesalvaro lead and installed Snap Camera to experiment with a different application and it did not exhibit the issue at all. I then went back to test again in Google Meet and the skew problem was gone. I'll be on the look out for this resurfacing and better prepared to identify any triggers. HTH.

Same exact situation on my side. Changing OBS resolution, restarting Chrome, restarting MacOS... none of these fixed the problem. Installing Snap Camera fixed the issue even when it isn't running.

**Comment by mdmarein** [comment]
Hello everyone, I had the same problem.
Here are the steps that I followed to correct it:
1. Close chrome
2. open OBS.
3. Verify that Virtual camera is off.
4. Go to Preferences in OBS> Video.
5. Base (canvas) and Output (sacled) resolution have to be the same.
6. Apply.
7. Close all, restart OSX

It worked for me.

**Comment by dmckinno** [comment]
I can confirm that @mdmarein's steps fixed the same issue for me as well. I'm pulling a video stream from a Sony a6300 through the Sony Remote app and streaming that window. 640x480 and 1280x720 both work for me.

@gxalpha, maybe the team automatically match base and output resolutions when virtual camera is triggered?

**Comment by wengzilla** [comment]
Confirmed that @mdmarein's steps fixed my issue as well...

**Comment by MarkusTiede** [comment]
I'm also affected by this issue - things I've noticed so far:
- @mdmarein steps don't work for me
- skewed output is present all the time: if OBS | virtual webcam is running or not
- initially it's been working exactly once

**Comment by KlausHB** [comment]
Hi!

in my setup i unchecked the deactivation of macOS v-sync and everything works now :-))

Klaus

**Comment by GuitarKat** [comment]
Changing the resolution to be a proper ratio to the input canvas really helped (this was OBS mac virtual cam for streamlabs). I would check there. It might have changed the resolution of the output when I ran the optimization on it.

**Comment by kushaldas** [comment]
> Very interesting I wonder if this is an obs bug or if I'm doing something wrong with converting pixel formats. @dodgepong have you seen anything like this in other issues on OBS? Seems like @tejasftc had the format set to NV12 even though his computer reported that it wasn't supported. This then seemed to break the pixel format conversion to yuv422 that my plugin does

I just now had to move to a different format to make it working. I chose RGB. MacOS `10.15.7` here with `1.2.0` release.

**Comment by PatTheMav** [comment]
Maybe a stupid question, but has this issue happened anywhere _but_ Google Meet? Because I tried all kinds of weird resolutions and also different aspect ratios between base and output resolution, etc. and a simple WebRTC page in Chrome and can't get the issue to appear.

Would be interesting if someone who _can_ reproduce the issue check if it also affects this test page at the same time: https://webrtc.github.io/samples/src/content/getusermedia/gum/

**Comment by gxalpha** [comment]
Seems like yes https://github.com/johnboiles/obs-mac-virtualcam/issues/192

**Comment by PatTheMav** [comment]
@gxalpha That looks like Google Meet again, or did I overlook something?

Anyway one thing I _was_ just able to find was that Chrome doesn't seem to "notice" that I changed the output resolution:

* Set Output to 1080p
* Set Canvas to any weird resolution
* Join Google Meet
* Observe cropped output
* Stop VirtualCam output
* Change canvas resolution to 1280x720
* Start VirtualCam output
* Join Google Meet
* Observe output is still cropped

Checking the output with Quicktime player at the same time showed the proper 720p image. So I don't know why Chrome doesn't pick up the new resolution even though the DAL plugin obviously did.

**Comment by gxalpha** [comment]
He said Google Meet and other platforms like that webcamtest.

Interesting find though! As far as I remember some apps need restarts after a resolution is changed (Skype if I remember correctly, that was the reason the test card got it's variable resolution), can't remember Chrome was one of them :thinking:

**Comment by PatTheMav** [comment]
Restarting Chrome makes it pick up the new resolution. Interestingly the issue appears even when OBS is not in the mix (as demonstrated by the test card being skewed as well).

Now the test card either picks up the 720p defaults or the last frame setup received from the Mach Server (OBS) as stored in the `NSUserDefaults` which would also explain why the issue persists across restarts and "overwriting" the plugin.

So whatever the DAL plugin received "broke" the output. If I'm not mistaken the file `~/Library/Preferences/com.obsproject.obs-studio.plist` seems to store the values for the unreleased built-in version, is there a similar file for the standalone plugin? Would be interesting to read what it says for the people that do _have_ the skewed output.

**Comment by pixelass** [comment]
@PatTheMav I got it fixed by following the comments/suggestions of #224 
Would the content of that file still help you?

**Comment by PatTheMav** [comment]
@pixelass Which suggestions fixed it for you?

**Comment by pixelass** [comment]
@PatTheMav sorry, I linked the wrong issue: It was THIS issue, specifically this comment: https://github.com/johnboiles/obs-mac-virtualcam/issues/160#issuecomment-671118838

**Comment by PatTheMav** [comment]
Thanks, yeah the weird thing is that I can't replicate the issue with all kinds of aspect ratios, resolutions and even different combinations of aspect ratios and resolutions between output and canvas resolutions. Without a Chrome restart the browser will always pick up the "last known" resolution, but the resulting image is then just cut appropriately while "native" apps pick up the new resolutions just fine.

**Comment by johnboiles** [comment]
@tejasftc @TayKangSheng @mgw-sbic @saipanman95 @wesalvaro @stevewww @truncj @mdmarein @dmckinno @wengzilla @MarkusTiede @KlausHB @GuitarKat @kushaldas @pixelass

Sorry for the wide ping but I need y'all's help to see if I've fixed this issue. I've never been able to reproduce it myself so I have no way of knowing if I've fixed it without your help. Here's what you can do to help:

1. Try to reproduce the issue with the version of the plugin you have installed (<=1.2.0). Try to do the opposite of whatever you did to resolve the solution (e.g. change resolutions to something weird, disable v-sync, etc). If you're able to reproduce the original skewed output issue, continue to the following steps.
2. Install the [latest v1.2.1 release](https://github.com/johnboiles/obs-mac-virtualcam/releases/tag/v1.2.1). It has a change that _hopefully_ fixes this bug.
3. Open up your host app of choice (e.g. Google Meet/Chrome, QuickTime, etc) and select OBS Virtual Camera as the source. Look at the blue test card to make sure it renders correctly. While you're there, double check that it says '1.2.1' at the bottom.
4. Open up OBS and start the virtual camera (Tools -> Start Virtual Camera), and make sure whatever is in your OBS renders correctly in your host app.

Let me know if you are able to repro and if my fix solved it! @PatTheMav is doing some great [work](https://github.com/obsproject/obs-studio/pull/3492) to get this all merged into the core OBS app, and it'd be great to have this hammered out before it's bundled by default with OBS. Thanks!

**Comment by KlausHB** [comment]
Hi John,

I am very busy this week. I check this out next week, if it helps You.
If others did the job, please let me know ;-)

CU

Klaus

**Comment by mikehatch** [comment]
I'm on v1.2.1 and had this issue - reboot didn't help, and changing the Video->Base Resolution & Output Resolution to be the same didn't help.  After reviewing comments and settings in this thread, the thing that fixed it was changing the setting: Advanced -> Video : Color Format from NV12 to something else (RGB) fixed it.  Then I changed it back to NV12 and it was fine.

**Comment by johnboiles** [comment]
@mikehatch thanks for the data point! Did you see the issue both in the test card and the video from OBS, or just one of those?

**Comment by mikehatch** [comment]
> @mikehatch thanks for the data point! Did you see the issue both in the test card and the video from OBS, or just one of those?

Yes, both the output and the test card were showing the skewed output as shown in this thread.

**Comment by PatTheMav** [comment]
So futzing around with the colour space might be another factor in triggering it? Will give it a whirl this weekend..

**Comment by johnboiles** [comment]
Yeah potentially anything that could change or recreate the pixel buffers on OBS or in the DAL plugin seems like a likely candidate.

**Comment by PatTheMav** [comment]
What I still try to wrap my head around is how even the test card will appear skewed, as reported by some people. The test card is created/rendered entirely within the DAL plugin with no input from OBS (where a different colour space would be set-up) and iirc only resolution and frame rate are stored in `UserDefaults`. Of course if the DAL plugin instance stays "alive" with bad settings, it won't recover, but I'm under the assumption that e.g. Chrome's plugin host runs its own instance of the DAL plugin and when Chrome is closed that one should be torn down as well. 🤔

**Comment by johnboiles** [comment]
Yeah it's very mysterious. It's a good point about the test card -- that's probably the place to start. It will be a more self contained problem to understand why the test card is skewed. And once we understand that maybe we can reason about why changing settings in OBS sometimes fixes it.

**Comment by johnboiles** [comment]
I'd start with looking at the [creation of the `CVPixelBuffer`](https://github.com/obsproject/obs-studio/blob/master/plugins/mac-virtualcam/src/dal-plugin/Stream.mm#L256). But without having a developer that can reproduce, all we can do is read and check the code. Hopefully your eyes will see something mine haven't.

**Comment by worst-Ness** [comment]
I'm having the same issue with OBS 1.2.1 on rockoutloud.live
![Uploading Screen Shot 2020-11-10 at 2.03.06 PM.png…]()

**Comment by TobiasBales** [comment]
I am having the same issues for some time now - anything I could provide to help you better understand the issue?

**Comment by pixelass** [comment]
There's not much to understand right now. Nobody really seems to know why this happens. Your best bet would be to follow the suggestions and if you notice a step that solved it for you, then mention it here so contributors can check for reproducibility.

**Comment by loretod** [comment]
Not sure why but updating to Chrome 87.0.4280.66 (Official Build) (x86_64) fixed it for me in Google Meet. it's been working great in Zoom so must have been a Chrome issue. Not sure if anything in the update log helps with the detective work? [https://chromereleases.googleblog.com/](https://chromereleases.googleblog.com/)

**Comment by capital-G** [comment]
@worst-Ness 
following settings worked for me with Jitsi and Chrome facing the same issue before

![grafik](https://user-images.githubusercontent.com/8267062/99690830-d73dc300-2a88-11eb-98ef-76c421913257.png)


![grafik](https://user-images.githubusercontent.com/8267062/99690801-d016b500-2a88-11eb-9c6c-d2994116cfb7.png)

**Comment by evangodfreysocal** [comment]
I also have this issue. Switching to Colour Format does resolve the tearing but the frame rate drop makes it unusable. With NV12 color format the frame rate is fine (though of course there's the tearing..)

I am reproing in Firefox/Chrome. I'm just googling webcam tests and picking from the top sites. Don't have Zoom or other meeting applications installed so haven't tested there. 

Plugin Version 1.2.1, used Package installer
OBS 26.0.2 (mac)

Specs (Yes I know):
MacBook Air 11" Mid-2013
macOS 11.0.1 Beta
1.3 GHz Dual-Core Intel Core i5
8 GB memory
Intel HD Graphics 5000 1636 MB

Please let me know if I can provide any additional information. I don't really need this for anything, it's just for fun, but happy to help if I can!

**Comment by joseluistubert** [comment]
I'm having the same problem. Tried every suggestion in the thread but nothing works. I still get the skewed image.

**Comment by joseluistubert** [comment]
The data card is also skewed.

**Comment by joseluistubert** [comment]
I installed this release obs-mac-virtualcam-4bd5852-v1.2.1.pkg
and that fixed the problem!!!! Thanks a lot John!!!!

**Comment by medwingADMIN** [comment]
This bug is not fixed for me unfortunately. Exactly the same issue
OS Version 11.0.1
Plugin 1.3.1 (I tried with 1.2.1 as well)
OBS - 26.02
The skewed video is there even when I have the virtual camera turned off and I have of course restarted etc

<img width="1317" alt="Screenshot 2020-11-27 at 12 20 46 pm" src="https://user-images.githubusercontent.com/51358138/100444015-0a94d900-30ab-11eb-96b7-60615c055c8a.png">

**Comment by johnboiles** [comment]
@medwingADMIN what are your resolution + color space + color format settings (see @capital-G's [post](https://github.com/johnboiles/obs-mac-virtualcam/issues/160#issuecomment-730472045) above)

**Comment by johnboiles** [comment]
I wonder if this could be the same underlying cause as #228 and #233 (attempted fix in #241). It seems that something goes wrong with sending and/or saving the FPS value from the OBS plugin to the DAL plugin. Totally possible the same thing could be happening for the width / height, which could potentially cause this issue.

**Comment by PatTheMav** [comment]
The "current" values for the plugin should be stored in the `plist` of the host application. So if this happens in Chrome, it should be set in `~/Library/Preferences/com.google.Chrome.helper.plugin.plist` or similar.

**Comment by johnboiles** [comment]
If this is a similar issue as #228 and #233, I added some logs in [`d11f17eba`](https://github.com/johnboiles/obs-mac-virtualcam/commit/d11f17eba39cff2fa5573846cdd81299aaeb24d9) that might help diagnose this. Go download the [pkg-installer](https://github.com/johnboiles/obs-mac-virtualcam/suites/1572984574/artifacts/28367683) and try it out while looking at the Console.app logs? https://github.com/johnboiles/obs-mac-virtualcam/runs/1465076120

You're looking for the logs that start with:
```
Saving frame info
Setting default resolution
Loaded resolution from NSUserDefaults
Setting default framerate
Loaded framerate from NSUserDefaults
```

**Comment by johnboiles** [comment]
@PatTheMav that's super helpful! Anyone who is experiencing this issue could you post your `plist` files (like the one @PatTheMav mentions). These should contain the saved defaults from the plugins.

**Comment by anthonypillot** [comment]
Posting my case too:

**Describe the bug**
Installed the plugin, the plugin worked fine on Google Chrome, Google Meet : I was able to seen my OBS diffusion on Google Meet.

But after a restart of Chrome, the stream is full blurred, skewed.

**To Reproduce**
Trying to reproduce the bug, reinstalling the plugin, restart my computer, full reinstall OBS, the plugin, but nothing works again.

**Expected behavior**
Just see my diffusion in Google Meet.

**Screenshots**

<img width="1383" alt="Capture d’écran 2020-11-23 à 17 21 17" src="https://user-images.githubusercontent.com/51130139/99987443-c48cfc00-2db0-11eb-8353-fbd5fcc4e871.png">

**Desktop (please complete the following information):**
 - OS version: [macOS Big Sur 11.0.1]
 - Application: [Chrome, Google Meet]
 - Virtualcam version: [1.2.1]
 - OBS version: [26.0.2]

**Comment by olifrieda** [comment]
**Having the same issue with:**
Color Format: NV12
Color Space: 709
Color Range: Partial
x Disable macOS V-Sync
x Reset macOS V-Sync on Exit

Base Resolution: 1792 x 1120
Output Resolution: 1192 x 746
Downscale Filter: Bicubic
Common FPS Value: 25 PAL

OS: 10.15.7
OBS: 26.0.2
Virtual Cam: 1.3.1
Google Chrome: 87.0.4280.88

**Solved it by:** (Don't forget to restart Google Chrome after changing the settings in OBS!)
Base Resolution: 1792 x 1120
Output Resolution: 1792 x 1120
(other resolutions are working too, but must be the same for both settings)

Scaled output resolution seems not to be working.
